BFD-4198: V3 SAMHSA filter implementation

aschey-forpeople opened the pull request:
<!--
You've got a Pull Request you want to submit? Awesome!
This PR template is here to help ensure you're setup for success:
  please fill it out to help ensure that your PR is complete and ready for approval.
-->

**JIRA Ticket:**
BFD-4198


### What Does This PR Do?
<!--
Add detailed description & discussion of changes here.
The contents of this section should be used as your commit message (unless you merge the PR via a merge commit, of course).

Please follow standard Git commit message guidelines:
* First line should be a capitalized, short (50 chars or less) summary.
* The rest of the message should be in standard Markdown format, wrapped to 72 characters.
* Describe your changes in imperative mood, e.g. "make xyzzy do frotz" instead of "[This patch] makes xyzzy do frotz" or "[I] changed xyzzy to do frotz", as if you are giving orders to the codebase to change its behavior.
* List all relevant Jira issue keys, one per line at the end of the message, per: <https://confluence.atlassian.com/jirasoftwarecloud/processing-issues-with-smart-commits-788960027.html>.

Reference: <https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project>.
-->

Adds SAMHSA filtering logic for EOB.

### What Should Reviewers Watch For?
<!--
Common items include:
* Is this likely to address the goals expressed in the user story?
* Are any additional documentation updates needed?
* Are there any unhandled and/or untested edge cases you can think of?
* Is user input properly sanitized & handled?
* Does this make any backwards-incompatible changes that might break end user clients?
* Can you find any bugs if you run the code locally and test it manually?
-->

If you're reviewing this PR, please check for these things in particular:
<!-- Add some items here -->

### What Security Implications Does This PR Have?

Please indicate if this PR does any of the following:  

* Adds any new software dependencies
* Modifies any security controls
* Adds new transmission or storage of data
* Any other changes that could possibly affect security? 

* [x] I have considered the above security implications as it relates to this PR. (If one or more of the above apply, it cannot be merged without the ISSO or team security engineer's (`@sb-benohe`) approval.) 
* [x] I have created tests to sufficiently ensure the reliability of my code, if applicable. If this is a modification to an existing piece of code, I have audited the associated tests to ensure everything works as expected.

### Validation

**Have you fully verified and tested these changes? Is the acceptance criteria met? Please provide reproducible testing instructions, code snippets, or screenshots as applicable.**
    
tests

aschey-forpeople commented on apps/bfd-server-ng/src/test/java/gov/cms/bfd/server/ng/EobSamhsaFilterIT.java:
```diff
@@ -0,0 +1,444 @@
+  private static final long CLAIM_UNIQUE_ID_FOR_FUTURE_HCPCS = 6871761612138L;
+  private static final long CLAIM_UNIQUE_ID_FOR_FUTURE_HCPCS_AFTER_CODE_START = 6871761612139L;
+
+  private static final long CLAIM_UNIQUE_ID_WITH_NO_SAMHSA = 566745788569L;
```
At some point I'd like to organize these test cases a bit better, but I think this will have to do for now.

brick-green commented on apps/bfd-server-ng/src/main/java/gov/cms/bfd/server/ng/eob/EobHandler.java:
```diff
@@ -71,18 +103,90 @@ public Bundle searchByBene(
+  private boolean procedureMatchesSamhsaCode(ClaimProcedure proc, LocalDate claimDate) {
+    var diagnosisCode = proc.getDiagnosisCode().orElse("");
+    var procedureCode = proc.getProcedureCode().orElse("");
+    var icdIndicator = proc.getIcdIndicator().orElse(IcdIndicator.ICD_10);
```
Add a comment saying why we default to ICD-10 for the future devs?

aschey-forpeople replied:
Added 👍

brick-green commented on apps/bfd-server-ng/src/main/java/gov/cms/bfd/server/ng/eob/EobHandler.java:
```diff
@@ -71,18 +103,90 @@ public Bundle searchByBene(
+  }
+
+  // Checks ICDs.
+  private boolean procedureMatchesSamhsaCode(ClaimProcedure proc, LocalDate claimDate) {
```
feels a little weird to name this `procedureMatchesSamhsaCode` then check procedure and diagnosis codes

brick-green commented on apps/bfd-server-ng/src/main/java/gov/cms/bfd/server/ng/eob/EobHandler.java:
```diff
@@ -71,18 +103,90 @@ public Bundle searchByBene(
+  }
+
+  private boolean hcpcsCodeMatches(ClaimLine claimLine, LocalDate claimDate) {
+    var hcpcs = claimLine.getHcpcsCode().getHcpcsCode().orElse("");
```
`getHcpcsCode().getHcpcsCode()` feels weird too. Out of scope here though

aschey-forpeople replied:
Yeah, it's because the field name is the same as the class name. We could maybe rename the field name to `code` instead of `hcpcsCode` but it does break our conventions a bit.

brick-green commented on apps/bfd-server-ng/src/main/java/gov/cms/bfd/server/ng/SecurityLabel.java:
```diff
@@ -0,0 +1,140 @@
+    try {
+      var securityLabels = new HashMap<String, List<SecurityLabel>>();
+      var yamlData =
+          new YAMLMapper().readValue(inputStream, new TypeReference<List<SecurityLabel>>() {});
```
Need to create a new YAMLMapper on every call here? Maybe pull out to a static?

brick-green reviewed (commented):
There were a few opportunities to short circuit but based on your response to Mahi I skipped over them

brick-green commented on apps/bfd-server-ng/src/main/java/gov/cms/bfd/server/ng/claim/model/Claim.java:
```diff
@@ -117,14 +127,16 @@ public ExplanationOfBenefit toFhir() {
-              .toFhirDiagnosis(item.getClaimItemId().getBfdRowId())
-              .ifPresent(eob::addDiagnosis);
-        });
+    claimItems.stream()
+        .sorted(Comparator.comparing(c -> c.getClaimItemId().getBfdRowId()))
+        .forEach(
+            item -> {
+              item.getClaimLine().toFhir(item).ifPresent(eob::addItem);
+              item.getClaimProcedure().toFhirProcedure().ifPresent(eob::addProcedure);
+              item.getClaimProcedure()
+                  .toFhirDiagnosis(item.getClaimItemId().getBfdRowId())
+                  .ifPresent(eob::addDiagnosis);
+            });
```
So that if it's called multiple times it will always return the same. Makes sense

np-complete0 commented on apps/bfd-pipeline/bfd-pipeline-idr/run-db.sh:
```diff
@@ -41,7 +41,7 @@ docker exec -u postgres bfd-idr-db psql idr bfd -f docker-entrypoint-initdb.d/mo
 
 uv sync
-BFD_DB_ENDPOINT=localhost BFD_DB_USERNAME=bfd BFD_DB_PASSWORD=InsecureLocalDev uv run python ./load_synthetic.py "$1"
+BFD_DB_ENDPOINT=localhost BFD_DB_USERNAME=bfd BFD_DB_PASSWORD=InsecureLocalDev uv run load_synthetic.py "$1" && uv run pipeline.py local
```
and username

np-complete0 commented on apps/bfd-server-ng/src/main/resources/security_labels.yml:
```diff
@@ -0,0 +1,2733 @@
+- system: http://www.ama-assn.org/go/cpt
```
do we have a third copy of this file in the project now ?

np-complete0 commented on apps/bfd-server-ng/src/main/resources/security_labels.yml:
```diff
@@ -0,0 +1,2733 @@
+- system: https://www.cms.gov/Medicare/Coding/HCPCSReleaseCodeSets
+  code: H0010
+  startDate: 1970-01-01
+  endDate: 9999-12-31
```
isn't that easier to filter out the ACTIVE ones instead of checking every entry and comparing if the year is less than the 9999 ?

np-complete0 commented:
I have added `cpt` code to `hcpcs` and api removed that (same as before). 
However, it does not honor the endDate. I changed endDate two day before today (2025-09-24), it was still removing whole entry ?

aschey-forpeople commented on apps/bfd-server-ng/src/main/java/gov/cms/bfd/server/ng/claim/model/Claim.java:
```diff
@@ -117,14 +127,16 @@ public ExplanationOfBenefit toFhir() {
-              .ifPresent(eob::addDiagnosis);
-        });
+    claimItems.stream()
+        .sorted(Comparator.comparing(c -> c.getClaimItemId().getBfdRowId()))
```
Yes, but the sequence numbers are generated based on the order that the claim lines get mapped, so we have to sort them first to make sure the sequence numbers match the data.

aschey-forpeople commented:
> I have added `cpt` code to `hcpcs` and api removed that (same as before). However, it does not honor the endDate. I changed endDate two day before today (2025-09-24), it was still removing whole entry ?

If you're talking about if the claim end date is before the start date, we decided to filter those on ingestion rather than accounting for that here.

aschey-forpeople commented on apps/bfd-model-idr/StructureMaps/.gitkeep:
This needs to be here because the FHIR validator CLI isn't smart enough to create missing folders. This is an issue left over from when I moved all the files around in here.

aschey-forpeople commented on apps/bfd-server-ng/src/test/java/gov/cms/bfd/server/ng/IntegrationTestBase.java:
```diff
@@ -62,6 +65,10 @@ protected IGenericClient getFhirClient() {
     return ctx.newRestfulGenericClient(getServerUrl());
   }
```
`FhirContext.forR4Cached()` uses a cached context if one is available (hence the name) so there's no need to make this static.

aschey-forpeople commented on apps/bfd-server-ng/src/main/java/gov/cms/bfd/server/ng/eob/EobResourceProvider.java:
We will address the SonarQube warnings as they come up in changed code, but I'm not going to go back and address a bunch of unrelated code in this PR. There's no need to copy/paste those warnings here.

aschey-forpeople commented on apps/bfd-server-ng/src/test/java/gov/cms/bfd/server/ng/EobSamhsaFilterIT.java:
```diff
@@ -0,0 +1,444 @@
+    for (var claimId : samhsaClaimIds) {
+      assertFalse(foundClaimIdsNoSamhsa.contains(claimId));
+      var read = eobRead().withId(claimId);
+      assertThrows(ResourceNotFoundException.class, read::execute);
```
I'm not sure I follow. The purpose of this is to assert that it throws `ResourceNotFoundException`.

aschey-forpeople commented on apps/bfd-server-ng/src/test/java/gov/cms/bfd/server/ng/EobSamhsaFilterIT.java:
```diff
@@ -0,0 +1,444 @@
+    var eobEnd = eob.getBillablePeriod().getEnd();
+    var json = normalize(context.newJsonParser().encodeResourceToString(eob));
+    for (var entry : SECURITY_LABELS.values().stream().flatMap(Collection::stream).toList()) {
+      if (validateDates && eobEnd.after(DateUtil.toDate(entry.getEndDateAsDate()))) {
```
Yeah I think these can be combined, fixed.

brick-green reviewed (approved):
:shipit:
